Detectors/EMCAL/simulation/src/DigitsWriteoutBuffer.cxx

```diff
@@ -20,62 +20,104 @@
+// Add digits to the buffer
+void DigitsWriteoutBuffer::addDigits(unsigned int towerID, std::vector<LabeledDigit> digList)
 {
-  int nsamples = int(eventTime / mTimeBinWidth);
 
-  if (nsamples >= mTimedDigits.size()) {
-    mTimedDigits.pop_front();
-    mTimedDigits.push_back(std::unordered_map<int, std::list<LabeledDigit>>());
-    nsamples = mTimedDigits.size() - 1;
-  }
+  for (int ientry = 0; ientry < mBufferSize; ientry++) {
 
-  auto& timeEntry = mTimedDigits[nsamples];
+    auto& buffEntry = mTimedDigitsFuture[ientry];
+    auto dig = digList.at(ientry);
 
-  auto towerEntry = timeEntry.find(towerID);
-  if (towerEntry == timeEntry.end()) {
-    towerEntry = timeEntry.insert(std::pair<int, std::list<o2::emcal::LabeledDigit>>(towerID, std::list<o2::emcal::LabeledDigit>())).first;
+    auto towerEntry = buffEntry.mDigitMap->find(towerID);
+    if (towerEntry == buffEntry.mDigitMap->end()) {
+      towerEntry = buffEntry.mDigitMap->insert(std::pair<int, std::list<o2::emcal::LabeledDigit>>(towerID, std::list<o2::emcal::LabeledDigit>())).first;
+    }
+
+    buffEntry.mTimestamp = dig.getTimeStamp();
+
+    // if the digit time is equal to the mLiveTime, the time bin is set to be the end time window
+    // if the digit time is between mLiveTime and busy time the time bin is set to not record mode
+    // if the digit time is > (mLiveTime + mBusyTime - mPreTriggerTime), the time bin will be set to record mode
+    if (dig.getTimeStamp() == mLiveTime) {
+      buffEntry.mRecordMode = false;
+      buffEntry.mEndWindow = true;
+    } else if (dig.getTimeStamp() >= mLiveTime && dig.getTimeStamp() < (mLiveTime + mBusyTime - mPreTriggerTime)) {
+      buffEntry.mRecordMode = false;
+      buffEntry.mEndWindow = false;
+    } else if (dig.getTimeStamp() < mLiveTime || dig.getTimeStamp() >= (mLiveTime + mBusyTime - mPreTriggerTime)) {
+      buffEntry.mRecordMode = true;
+      buffEntry.mEndWindow = false;
+    }
+
+    towerEntry->second.push_back(dig);
   }
+}
 
-  towerEntry->second.push_back(dig);
+// When the current time is forwarded (every 100 ns) the entry 0 of the future dequeue
+// is removed and pushed at the end of the past dequeue. At the same time a new entry in
+// the future dequeue is pushed at the end, and - in case the past dequeue reached 15 entries
+// the first entry of the past dequeue is removed.
+void DigitsWriteoutBuffer::forwardMarker(double eventTime)
+{
+
+  int sampleDifference = int(eventTime / 100) - int(mLastEventTime / 100);
+
+  for (int idel = 0; idel < sampleDifference; idel++) {
+    if (mTimedDigitsFuture.front().mEndWindow || !mTimedDigitsFuture.front().mRecordMode) {
+      break;
+    }
+
+    mTimedDigitsPast.push_back(mTimedDigitsFuture.front());
+    mTimedDigitsFuture.pop_front();
+    mTimedDigitsFuture.push_back(DigitTimebin());
+
+    if (mTimedDigitsPast.size() > mBufferSize) {
+      mTimedDigitsPast.pop_front();
+    }
```

Comment:
I was thinking to do the filling of the output vector here, because basically in this step, in case the last digit in the past dequeue has the close marker set, we know we have to stream the digits which are just created. The filling part would then be removed from the DigitizerSpec, which would simplify the Spec even more. Furthermore, it makes the object to what it is actually supposed to, indicated by its name: Writing out the digits created by the time sampling. One could do this in two ways, I have no strong preference here:

- Create a dedicated stream class which contains the digits vector, trigger records vector and MCLabelContainer, pass the entire std::dequeue to this
- Reference the 3 containers mentioned above - which are owned by the Spec and fill them inside the DigitsWriteoutBuffer in a dedicated fill function.

The first option would maybe have the small advantage that it itself the 3 containers, so it is self-contained.

Reply:
Done